src/libvirtApi/domain.js

```diff
@@ -1112,3 +1113,21 @@ export function domainUpdateDiskAttributes({ connectionName, objPath, target, re
                 return call(connectionName, '/org/libvirt/QEMU', 'org.libvirt.Connect', 'DomainDefineXML', [updatedXML], { timeout, type: 's' });
             });
 }
+
+export async function domainReplaceSpice({ connectionName, id: objPath }) {
+    /* Ideally this would be done by virt-xml, but it doesn't offer that functionality yet
+     * see https://issues.redhat.com/browse/RHEL-17436 */
+    const [domXML] = await call(connectionName, objPath, 'org.libvirt.Domain', 'GetXMLDesc', [Enum.VIR_DOMAIN_XML_INACTIVE], { timeout, type: 'u' });
+    const updatedXML = replaceSpice(domXML);
+
+    // check that updatedXML is valid; if not, throw; it needs to be updated manually
+    await (cockpit.spawn(["virt-xml-validate", "-"], { err: "message" }).input(updatedXML));
```

Comment:
Is there any advantage to doing a explicit validation step before calling "DomainDefineXML"? Wouldn't "DomainDefineXML" do the validation again and return the same error?

Reply:
Good point -- @crobinso @germanovm WDYT? Is this redundant/equivalent, or is virt-xml-validate stronger than what DomainDefineXML does? Thanks!

Reply:
For history: I originally did that to get a safety check before undefining the VM. But since then I took that out and just redefine (which is more atomic). So indeed I figure this is redundant now, but I'll rather wait for confirmation from the libvirt devs.

Reply:
I.e. I tried `virsh destroy test1`, edit ./.config/libvirt/qemu/test1.xml to add a `<foo bar='on'/>` into `<graphics>`. But `virsh start test1` works fine, and `virsh dumpxml --inactive test1` doesn't show the extra `<foo>` tag. I tried the same with adding an unknown attribute to `<redirdev bus='usb' type='spicevmc'>`, same effect -- unknown stuff gets ignored.

And when I replace e.g. `bus='usb'` to `bus='usc'` and try to start test1, then the VM disappears from libvirt completely.

Reply:
editing XML behind libvirt's back is tough. there's errors which still will load the XML and only fail at `virsh start` time, but a bad enum value like `usc` will cause it to completely fail to parse and disappear like you say.

maybe start with some spice specific very obscure config like '<channel name='cursor' mode='secure'/> in the spice `<graphics>` block. That bit isn't valid for VNC and the XML validator will reject it after s/spice/vnc/

different tactic: add `<gl enable='yes' rendernode='/fake/rendernode/path/make/cockpit/fail'/>` and then change cockpit code to match on that fake path and trigger error in some way. that XML will parse but the VM won't run obviously

Reply:
Ah, @crobinso I didn't see your most recent reply yet when I wrote back. That `<channel name='cursor' mode='secure'/>` sounds like an excellent way -- it wouldn't even be a hack, but something that could *actually* occur on a real system. However, when I do that, defining the updated XML just works, and libvirt automatically cleans up the XML to

```xml
   <graphics type='vnc' port='-1' autoport='yes' listen='127.0.0.1'>
      <listen type='address' address='127.0.0.1'/>
   </graphics>
```
At the same time, virt-xml-validate seems to be happy with the vnc + cursor/secure thing as well. So it seems this is actually quite unbreakable :grin: 

So I think I'll actually stand by my comment above -- I'm happy to tweak this further in the future, but it feels "good enough"?

Reply:
> I'm happy to tweak this further in the future, but it feels "good enough"?

Agreed. Covering the error case in the test is important and running virt-xml-validate is not wrong, and since it does a different kind of validation than DefineXML, it might even add value.

Reply:
Ack, closing this thread then. @crobinso thanks a lot for your help! :hugs: